common/transport/mqtt/src/mqtt_base.ts

```diff
@@ -253,39 +272,57 @@ export class MqttBase extends EventEmitter {
-    this._mqttClient.on('error', errorCallback);
-    this._mqttClient.on('close', closeCallback);
-    this._mqttClient.on('offline', offlineCallback);
-    this._mqttClient.on('disconnect', disconnectCallback);
+    this._mqttClient.on('message', this._messageEventFsmHandler);
+    this._mqttClient.on('error', connectingErrorCallback);
+    this._mqttClient.on('close', connectingCloseCallback);
+    this._mqttClient.on('offline', connectingOfflineCallback);
+    this._mqttClient.on('disconnect', connectingDisconnectCallback);
 
-    this._mqttClient.on('connect', (connack) => {
+    this._mqttClient.once('connect', (connack) => {
       debug('Device is connected');
       debug('CONNACK: ' + JSON.stringify(connack));
 
-      this._mqttClient.removeListener('close', closeCallback);
-      this._mqttClient.removeListener('offline', offlineCallback);
-      this._mqttClient.removeListener('disconnect', disconnectCallback);
+      this._mqttClient.on('error', this._errorEventFsmHandler);
+      this._mqttClient.on('close', this._closeEventFsmHandler);
+      this._mqttClient.on('offline', this._closeEventFsmHandler);
+      this._mqttClient.on('disconnect', this._closeEventFsmHandler);
+      this._mqttClient.removeListener('close', connectingCloseCallback);
+      this._mqttClient.removeListener('offline', connectingOfflineCallback);
+      this._mqttClient.removeListener('disconnect', connectingDisconnectCallback);
+      this._mqttClient.removeListener('error', connectingErrorCallback);
 
       callback(null, connack);
     });
   }
 
   private _disconnectClient(forceDisconnect: boolean, callback: () => void): void {
-    this._mqttClient.removeAllListeners();
     /* Codes_SRS_NODE_COMMON_MQTT_BASE_16_001: [The disconnect method shall call the done callback when the connection to the server has been closed.] */
-    this._mqttClient.end(forceDisconnect, callback);
+    let oldMqttClient = this._mqttClient;
+    this._mqttClient = undefined;
+
+    const oldMqttClientErrorHandler = (err) => debug('BUG? received an error on an MQTT client that is already being closed, ignoring it: ' + err.toString());
+    oldMqttClient.on('error', oldMqttClientErrorHandler);
+
+    // close is the event that indicates the end of the TCP connection, no chance to get a packet or an error after that.
+    oldMqttClient.once('close', () => {
+      debug('old MQTT client closed - removing the last error handler');
+      oldMqttClient.removeListener('error', oldMqttClientErrorHandler);
+      oldMqttClient = undefined;
+    });
+
+    oldMqttClient.removeListener('disconnect', this._closeEventFsmHandler);
+    oldMqttClient.removeListener('offline', this._closeEventFsmHandler);
+    oldMqttClient.removeListener('close', this._closeEventFsmHandler);
+    oldMqttClient.removeListener('message', this._messageEventFsmHandler);
+    oldMqttClient.removeListener('error', this._errorEventFsmHandler);
+
+    oldMqttClient.end(forceDisconnect, callback);
```

Comment:
>forceDisconnect [](start = 22, length = 15)

So if forceDisconnect is false, one might never see the close event?

Reply:
I believe we probably already got a close event if that were the case. forceDisconnect is used when the socket has already failed.

---
In reply to: [206239361](https://github.com/Azure/azure-iot-sdk-node/pull/343#discussion_r206239361) [](ancestors = 206239361)